src/lib.rs

```diff
@@ -651,26 +651,29 @@ impl<T> Receiver<T> {
     /// Panics if called after this receiver has been polled asynchronously.
     #[cfg(feature = "std")]
     pub fn recv_deadline(&self, deadline: Instant) -> Result<T, RecvTimeoutError> {
+        /// # Safety
+        ///
+        /// The state must be EMPTY (in this case that means UNPARKING) or MESSAGE
+        /// when calling this function. The message must also already have been written to
+        /// the channel and an acquire memory barrier issued to synchronize with that write.
         #[cold]
-        fn wait_for_unpark<T>(channel: &Channel<T>) -> Result<T, RecvTimeoutError> {
+        unsafe fn wait_for_unpark<T>(channel: &Channel<T>) -> Result<T, RecvTimeoutError> {
+            // We have observed the sender setting the UNPARKING state, and we swapped
+            // to the EMPTY state. The state is guaranteed to be EMPTY until the sender
+            // sets it to MESSAGE. No other states are possible here.
             loop {
                 thread::park();
-
-                // Same ordering as usual
-                match channel.state.load(Acquire) {
-                    MESSAGE => {
-                        // Same ordering and safety as usual
-
-                        channel.state.store(DISCONNECTED, Relaxed);
```

Comment:
Do you agree with me that we have already synchronized with the message store when we enter this function? If so, the acquire loads in this function should not be needed. We just wait until we observe the MESSAGE state.

I replaced your `load() + store()` with a `compare_exchange()`. Do you have any comments on this way of doing it?

Reply:
Fixed. Yeah, that makes sense. Plus, the `compare_exchange` did not make sense any longer when I had to handle the possible `DISCONNECTED` state also.